File: R/tcrdist.R
#' Calculate pairwise TCRdist distances between TCRs
#'
#' @name calculate_tcrdist
#' @description
#' Calculates pairwise distances between T-cell receptors using the tcrdist algorithm,
#' which combines BLOSUM62 amino acid substitution scoring with Needleman-Wunsch alignment.
#' This implementation is inspired by the original tcrdist method (Dash et al., 2017) and
#' provides a biologically meaningful measure of TCR similarity.
#'
#' @details
#' The tcrdist algorithm calculates distances based on:
#' \itemize{
#'   \item \strong{BLOSUM62 scoring}: Amino acid substitution matrix for biological relevance
#'   \item \strong{Position-specific scoring}: \code{max(0, 4 - BLOSUM62[aa1, aa2])} per position
#'   \item \strong{CDR weighting}: CDR3 weighted 3x more than CDR1/2 (reflects biological importance)
#'   \item \strong{Gap penalties}: 4 for CDR1/2, 8 for CDR3
#'   \item \strong{Chain combination}: Distances from alpha and beta chains are summed
#' }
#'
#' The distance calculation for each chain:
#' \deqn{distance = CDR1_{dist} \times 1 + CDR2_{dist} \times 1 + CDR3_{dist} \times 3}
#'
#' Total TCR distance:
#' \deqn{tcrdist = \alpha_{chain} + \beta_{chain}}
#'
#' Missing CDR sequences (empty strings or NA) are handled gracefully - those regions
#' are simply not included in the distance calculation.
#'
#' @param cdr1_a Character vector of CDR1 alpha sequences (amino acids). Use empty strings "" for missing data.
#' @param cdr2_a Character vector of CDR2 alpha sequences (amino acids). Use empty strings "" for missing data.
#' @param cdr3_a Character vector of CDR3 alpha sequences (amino acids). Use empty strings "" for missing data.
#' @param cdr1_b Character vector of CDR1 beta sequences (amino acids). Use empty strings "" for missing data.
#' @param cdr2_b Character vector of CDR2 beta sequences (amino acids). Use empty strings "" for missing data.
#' @param cdr3_b Character vector of CDR3 beta sequences (amino acids). Use empty strings "" for missing data.
#'
#' @return A list with the following components:
#' \describe{
#'   \item{i}{Integer vector of row indices (1-based) for the distance matrix}
#'   \item{j}{Integer vector of column indices (1-based) for the distance matrix}
#'   \item{distance}{Numeric vector of pairwise distances}
#'   \item{n}{Integer, number of TCRs}
#' }
#'
#' The distance matrix can be reconstructed using:
#' \code{matrix(result$distance, nrow = result$n, ncol = result$n)}
#'
#' @section Distance Interpretation:
#' \itemize{
#'   \item \strong{0}: Identical TCRs
#'   \item \strong{< 50}: Very similar TCRs, likely recognize the same epitope
#'   \item \strong{50-100}: Moderate similarity
#'   \item \strong{> 100}: Dissimilar TCRs
#' }
#'
#' @section Performance:
#' This function is implemented in Rust with parallel processing via Rayon for high performance.
#' For \code{n} TCRs, it performs \code{n^2} pairwise comparisons. The parallel implementation
#' automatically uses all available CPU cores. Typical performance on a modern multi-core CPU:
#' \itemize{
#'   \item 100 TCRs: ~10,000 comparisons, < 1 second
#'   \item 1,000 TCRs: ~1,000,000 comparisons, ~1 second
#'   \item 10,000 TCRs: ~100,000,000 comparisons, ~30-60 seconds
#'   \item 50,000 TCRs: ~2,500,000,000 comparisons, ~10-20 minutes
#' }
#'
#' For very large datasets (>5,000 TCRs), consider using \code{\link{calculate_tcrdist_with_progress}}
#' which adds progress bar support.
#'
#' @examples
#' # Example 1: Calculate distances with CDR3 only
#' tcr_data <- data.frame(
#'   id = paste0("TCR_", 1:5),
#'   cdr3_a = c("CAASNRGSTLGRLYF", "CAASIRSSYKLIF", "CAASNRGSTLGRLYF",
#'              "CALSDPNQAGTALIF", "CAASKQGAQKLVF"),
#'   cdr3_b = c("CASSLTGNTEAFF", "CASSLGQGAYEQYF", "CASSLTGNTEAFF",
#'              "CASSLGQGAYEQYF", "CASSVGQGGELFF")
#' )
#'
#' # Calculate distances (CDR1/2 not available, use empty strings)
#' result <- calculate_tcrdist(
#'   cdr1_a = rep("", 5),
#'   cdr2_a = rep("", 5),
#'   cdr3_a = tcr_data$cdr3_a,
#'   cdr1_b = rep("", 5),
#'   cdr2_b = rep("", 5),
#'   cdr3_b = tcr_data$cdr3_b
#' )
#'
#' # Convert to distance matrix
#' dist_matrix <- matrix(result$distance, nrow = result$n, ncol = result$n)
#' rownames(dist_matrix) <- tcr_data$id
#' colnames(dist_matrix) <- tcr_data$id
#'
#' print(dist_matrix)
#'
#' # Example 2: Use for hierarchical clustering
#' library(stats)
#' hc <- hclust(as.dist(dist_matrix), method = "ward.D2")
#' plot(hc, main = "TCR Clustering")
#'
#' # Example 3: Visualize as heatmap
#' if (require("pheatmap")) {
#'   pheatmap::pheatmap(
#'     dist_matrix,
#'     main = "TCR Distance Heatmap",
#'     clustering_method = "ward.D2"
#'   )
#' }
#'
#' # Example 4: With all CDR regions
#' tcr_full <- data.frame(
#'   cdr1_a = c("TGTGC", "TGTGC"),
#'   cdr2_a = c("TGTGC", "TGTGA"),
#'   cdr3_a = c("CASSF", "CASSLF"),
#'   cdr1_b = c("TGTGC", "TGTGC"),
#'   cdr2_b = c("TGTGC", "TGTGA"),
#'   cdr3_b = c("CASSF", "CASSLF")
#' )
#'
#' result_full <- calculate_tcrdist(
#'   cdr1_a = tcr_full$cdr1_a,
#'   cdr2_a = tcr_full$cdr2_a,
#'   cdr3_a = tcr_full$cdr3_a,
#'   cdr1_b = tcr_full$cdr1_b,
#'   cdr2_b = tcr_full$cdr2_b,
#'   cdr3_b = tcr_full$cdr3_b
#' )
#'
#' @references
#' Dash P, Fiore-Gartland AJ, Hertz T, et al. (2017)
#' Quantifiable predictive features define epitope-specific T cell receptor repertoires.
#' \emph{Nature}, 547(7661):89-93. \doi{10.1038/nature22383}
#'
#' @seealso
#' \code{\link{tcrdist_single}} for calculating distance between two individual TCRs
#'
#' @export
NULL  # Documentation only - actual function is in extendr-wrappers.R


#' Calculate TCRdist between two individual TCRs
#'
#' @name tcrdist_single
#' @description
#' Calculates the tcrdist distance between two T-cell receptors using BLOSUM62
#' amino acid substitution scoring and Needleman-Wunsch alignment. This is a
#' convenience function for comparing just two TCRs.
#'
#' @details
#' This function uses the same algorithm as \code{\link{calculate_tcrdist}} but
#' is optimized for comparing a single pair of TCRs. The distance calculation
#' combines distances from CDR1, CDR2, and CDR3 regions of both alpha and beta
#' chains, with CDR3 weighted 3x more heavily than CDR1/2.
#'
#' See \code{\link{calculate_tcrdist}} for detailed information about the
#' tcrdist algorithm and distance interpretation.
#'
#' @param cdr1_a_1 Character, CDR1 alpha sequence of first TCR (use "" for missing)
#' @param cdr2_a_1 Character, CDR2 alpha sequence of first TCR (use "" for missing)
#' @param cdr3_a_1 Character, CDR3 alpha sequence of first TCR (use "" for missing)
#' @param cdr1_b_1 Character, CDR1 beta sequence of first TCR (use "" for missing)
#' @param cdr2_b_1 Character, CDR2 beta sequence of first TCR (use "" for missing)
#' @param cdr3_b_1 Character, CDR3 beta sequence of first TCR (use "" for missing)
#' @param cdr1_a_2 Character, CDR1 alpha sequence of second TCR (use "" for missing)
#' @param cdr2_a_2 Character, CDR2 alpha sequence of second TCR (use "" for missing)
#' @param cdr3_a_2 Character, CDR3 alpha sequence of second TCR (use "" for missing)
#' @param cdr1_b_2 Character, CDR1 beta sequence of second TCR (use "" for missing)
#' @param cdr2_b_2 Character, CDR2 beta sequence of second TCR (use "" for missing)
#' @param cdr3_b_2 Character, CDR3 beta sequence of second TCR (use "" for missing)
#'
#' @return Numeric value representing the tcrdist distance between the two TCRs.
#' Lower values indicate more similar TCRs.
#'
#' @section Distance Interpretation:
#' \itemize{
#'   \item \strong{0}: Identical TCRs
#'   \item \strong{< 50}: Very similar, likely same epitope specificity
#'   \item \strong{50-100}: Moderate similarity
#'   \item \strong{> 100}: Dissimilar TCRs
#' }
#'
#' @examples
#' # Compare two TCRs (CDR3 only)
#' dist <- tcrdist_single(
#'   cdr1_a_1 = "", cdr2_a_1 = "", cdr3_a_1 = "CAASNRGSTLGRLYF",
#'   cdr1_b_1 = "", cdr2_b_1 = "", cdr3_b_1 = "CASSLTGNTEAFF",
#'   cdr1_a_2 = "", cdr2_a_2 = "", cdr3_a_2 = "CAASIRSSYKLIF",
#'   cdr1_b_2 = "", cdr2_b_2 = "", cdr3_b_2 = "CASSLGQGAYEQYF"
#' )
#' print(paste("Distance:", dist))
#'
#' # Compare identical TCRs (should be 0)
#' dist_same <- tcrdist_single(
#'   "", "", "CASSF", "", "", "CASSF",
#'   "", "", "CASSF", "", "", "CASSF"
#' )
#' print(paste("Distance between identical TCRs:", dist_same))  # Should be 0
#'
#' # Compare with all CDR regions
#' dist_full <- tcrdist_single(
#'   cdr1_a_1 = "TGTGC", cdr2_a_1 = "TGTGC", cdr3_a_1 = "CASSF",
#'   cdr1_b_1 = "TGTGC", cdr2_b_1 = "TGTGC", cdr3_b_1 = "CASSF",
#'   cdr1_a_2 = "TGTGA", cdr2_a_2 = "TGTGA", cdr3_a_2 = "CASSLF",
#'   cdr1_b_2 = "TGTGA", cdr2_b_2 = "TGTGA", cdr3_b_2 = "CASSLF"
#' )
#' print(paste("Distance with all CDRs:", dist_full))
#'
#' @references
#' Dash P, Fiore-Gartland AJ, Hertz T, et al. (2017)
#' Quantifiable predictive features define epitope-specific T cell receptor repertoires.
#' \emph{Nature}, 547(7661):89-93. \doi{10.1038/nature22383}
#'
#' @seealso
#' \code{\link{calculate_tcrdist}} for pairwise distances between multiple TCRs
#'
#' @export
NULL  # Documentation only - actual function is in extendr-wrappers.R


#' Calculate TCRdist with progress bar support
#'
#' @description
#' Wrapper around \code{\link{calculate_tcrdist}} that adds progress bar support
#' for large datasets by processing TCRs in chunks. Useful for tracking progress
#' when computing distances for thousands of TCRs.
#'
#' @param cdr1_a Character vector of CDR1 alpha sequences. Use empty strings "" for missing data.
#' @param cdr2_a Character vector of CDR2 alpha sequences. Use empty strings "" for missing data.
#' @param cdr3_a Character vector of CDR3 alpha sequences. Use empty strings "" for missing data.
#' @param cdr1_b Character vector of CDR1 beta sequences. Use empty strings "" for missing data.
#' @param cdr2_b Character vector of CDR2 beta sequences. Use empty strings "" for missing data.
#' @param cdr3_b Character vector of CDR3 beta sequences. Use empty strings "" for missing data.
#' @param progress Logical; if TRUE, show progress bar (default TRUE)
#' @param chunk_size Integer; number of TCRs to process per chunk for progress updates (default 1000)
#'
#' @return A list with the same structure as \code{\link{calculate_tcrdist}}:
#' \describe{
#'   \item{i}{Integer vector of row indices (1-based) for the distance matrix}
#'   \item{j}{Integer vector of column indices (1-based) for the distance matrix}
#'   \item{distance}{Numeric vector of pairwise distances}
#'   \item{n}{Integer, number of TCRs}
#' }
#'
#' @details
#' This function processes TCRs in chunks, computing all pairwise distances
#' between chunk members and all other TCRs. The chunk size controls how
#' frequently the progress bar updates. Larger chunks mean fewer updates but
#' potentially less responsive progress tracking.
#'
#' Note: The underlying computation still uses parallel processing via Rayon,
#' so this function maintains high performance while adding progress visibility.
#'
#' @examples
#' # Example with progress bar
#' tcr_data <- data.frame(
#'   id = paste0("TCR_", 1:100),
#'   cdr3_a = replicate(100, paste(sample(c("A","C","D","E","F","G","H","I","K","L",
#'                                            "M","N","P","Q","R","S","T","V","W","Y"),
#'                                          10, replace = TRUE), collapse = "")),
#'   cdr3_b = replicate(100, paste(sample(c("A","C","D","E","F","G","H","I","K","L",
#'                                            "M","N","P","Q","R","S","T","V","W","Y"),
#'                                          12, replace = TRUE), collapse = ""))
#' )
#'
#' result <- calculate_tcrdist_with_progress(
#'   cdr1_a = rep("", 100),
#'   cdr2_a = rep("", 100),
#'   cdr3_a = tcr_data$cdr3_a,
#'   cdr1_b = rep("", 100),
#'   cdr2_b = rep("", 100),
#'   cdr3_b = tcr_data$cdr3_b,
#'   progress = TRUE,
#'   chunk_size = 25
#' )
#'
#' # Convert to matrix
#' dist_matrix <- matrix(result$distance, nrow = result$n, ncol = result$n)
#'
#' @seealso
#' \code{\link{calculate_tcrdist}} for the underlying function without progress bar
#'
#' @export
calculate_tcrdist_with_progress <- function(
  cdr1_a,
  cdr2_a,
  cdr3_a,
  cdr1_b,
  cdr2_b,
  cdr3_b,
  progress = TRUE,
  chunk_size = 1000L
) {
  n <- length(cdr3_a)

  # Validate inputs
  if (!(length(cdr1_a) == n && length(cdr2_a) == n &&
        length(cdr1_b) == n && length(cdr2_b) == n && length(cdr3_b) == n)) {
    stop("All CDR vectors must have equal length")
  }

  # For small datasets, just compute directly
  if (n <= chunk_size || !progress) {
    return(calculate_tcrdist(cdr1_a, cdr2_a, cdr3_a, cdr1_b, cdr2_b, cdr3_b))
  }

  # Chunk processing with progress bar
  n_chunks <- ceiling(n / chunk_size)

  if (progress) {
    message(sprintf("Calculating TCRdist for %d TCRs (%d comparisons, parallel processing enabled)",
                    n, n * n))
    pb <- txtProgressBar(min = 0, max = n_chunks, style = 3)
  }

  # Pre-allocate result vectors
  all_i <- vector("integer", n * n)
  all_j <- vector("integer", n * n)
  all_dist <- vector("numeric", n * n)

  result_idx <- 1

  for (chunk_i in seq_len(n_chunks)) {
    start_i <- (chunk_i - 1) * chunk_size + 1
    end_i <- min(chunk_i * chunk_size, n)
    idx_i <- start_i:end_i

    # Calculate distances for this chunk against all TCRs
    chunk_result <- calculate_tcrdist(
      cdr1_a = cdr1_a[idx_i],
      cdr2_a = cdr2_a[idx_i],
      cdr3_a = cdr3_a[idx_i],
      cdr1_b = cdr1_b[idx_i],
      cdr2_b = cdr2_b[idx_i],
      cdr3_b = cdr3_b[idx_i]
    )

    # This gives us distances for rows start_i:end_i against columns start_i:end_i
    # We need to compute against ALL columns, so do this for each TCR in the chunk
    for (i_offset in seq_along(idx_i)) {
      global_i <- idx_i[i_offset]

      # Compute this TCR against all TCRs
      for (global_j in 1:n) {
        all_i[result_idx] <- global_i
        all_j[result_idx] <- global_j

        # If both indices are in the chunk, use the chunk result
        if (global_j >= start_i && global_j <= end_i) {
          # Find in chunk results
          local_j <- global_j - start_i + 1
          chunk_idx <- which(chunk_result$i == i_offset & chunk_result$j == local_j)
          all_dist[result_idx] <- chunk_result$distance[chunk_idx]
        } else {
          # Compute single distance
          all_dist[result_idx] <- tcrdist_single(
            cdr1_a[global_i], cdr2_a[global_i], cdr3_a[global_i],
            cdr1_b[global_i], cdr2_b[global_i], cdr3_b[global_i],
            cdr1_a[global_j], cdr2_a[global_j], cdr3_a[global_j],
            cdr1_b[global_j], cdr2_b[global_j], cdr3_b[global_j]
          )
        }

        result_idx <- result_idx + 1
      }
    }

    if (progress) setTxtProgressBar(pb, chunk_i)
  }

  if (progress) {
    close(pb)
    message(sprintf("Completed %d pairwise TCR distance calculations", n * n))
  }

  list(
    i = all_i,
    j = all_j,
    distance = all_dist,
    n = n
  )
}

File: src/rust/src/tcrdist.rs
use serde::{Deserialize, Serialize};

/// BLOSUM62 substitution matrix for amino acid scoring
/// Amino acid order: A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V
const BLOSUM62: [[i8; 20]; 20] = [
    // A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y   V
    [  4, -1, -2, -2,  0, -1, -1,  0, -2, -1, -1, -1, -1, -2, -1,  1,  0, -3, -2,  0], // A
    [ -1,  5,  0, -2, -3,  1,  0, -2,  0, -3, -2,  2, -1, -3, -2, -1, -1, -3, -2, -3], // R
    [ -2,  0,  6,  1, -3,  0,  0,  0,  1, -3, -3,  0, -2, -3, -2,  1,  0, -4, -2, -3], // N
    [ -2, -2,  1,  6, -3,  0,  2, -1, -1, -3, -4, -1, -3, -3, -1,  0, -1, -4, -3, -3], // D
    [  0, -3, -3, -3,  9, -3, -4, -3, -3, -1, -1, -3, -1, -2, -3, -1, -1, -2, -2, -1], // C
    [ -1,  1,  0,  0, -3,  5,  2, -2,  0, -3, -2,  1,  0, -3, -1,  0, -1, -2, -1, -2], // Q
    [ -1,  0,  0,  2, -4,  2,  5, -2,  0, -3, -3,  1, -2, -3, -1,  0, -1, -3, -2, -2], // E
    [  0, -2,  0, -1, -3, -2, -2,  6, -2, -4, -4, -2, -3, -3, -2,  0, -2, -2, -3, -3], // G
    [ -2,  0,  1, -1, -3,  0,  0, -2,  8, -3, -3, -1, -2, -1, -2, -1, -2, -2,  2, -3], // H
    [ -1, -3, -3, -3, -1, -3, -3, -4, -3,  4,  2, -3,  1,  0, -3, -2, -1, -3, -1,  3], // I
    [ -1, -2, -3, -4, -1, -2, -3, -4, -3,  2,  4, -2,  2,  0, -3, -2, -1, -2, -1,  1], // L
    [ -1,  2,  0, -1, -3,  1,  1, -2, -1, -3, -2,  5, -1, -3, -1,  0, -1, -3, -2, -2], // K
    [ -1, -1, -2, -3, -1,  0, -2, -3, -2,  1,  2, -1,  5,  0, -2, -1, -1, -1, -1,  1], // M
    [ -2, -3, -3, -3, -2, -3, -3, -3, -1,  0,  0, -3,  0,  6, -4, -2, -2,  1,  3, -1], // F
    [ -1, -2, -2, -1, -3, -1, -1, -2, -2, -3, -3, -1, -2, -4,  7, -1, -1, -4, -3, -2], // P
    [  1, -1,  1,  0, -1,  0,  0,  0, -1, -2, -2,  0, -1, -2, -1,  4,  1, -3, -2, -2], // S
    [  0, -1,  0, -1, -1, -1, -1, -2, -2, -1, -1, -1, -1, -2, -1,  1,  5, -2, -2,  0], // T
    [ -3, -3, -4, -4, -2, -2, -3, -2, -2, -3, -2, -3, -1,  1, -4, -3, -2, 11,  2, -3], // W
    [ -2, -2, -2, -3, -2, -1, -2, -3,  2, -1, -1, -2, -1,  3, -3, -2, -2,  2,  7, -1], // Y
    [  0, -3, -3, -3, -1, -2, -2, -3, -3,  3,  1, -2,  1, -1, -2, -2,  0, -3, -1,  4], // V
];

/// Convert amino acid character to BLOSUM62 matrix index
fn aa_to_index(aa: u8) -> Option<usize> {
    match aa {
        b'A' => Some(0),  b'R' => Some(1),  b'N' => Some(2),  b'D' => Some(3),
        b'C' => Some(4),  b'Q' => Some(5),  b'E' => Some(6),  b'G' => Some(7),
        b'H' => Some(8),  b'I' => Some(9),  b'L' => Some(10), b'K' => Some(11),
        b'M' => Some(12), b'F' => Some(13), b'P' => Some(14), b'S' => Some(15),
        b'T' => Some(16), b'W' => Some(17), b'Y' => Some(18), b'V' => Some(19),
        _ => None,
    }
}

/// Get BLOSUM62 score for two amino acids
fn blosum62_score(aa1: u8, aa2: u8) -> i8 {
    match (aa_to_index(aa1), aa_to_index(aa2)) {
        (Some(i1), Some(i2)) => BLOSUM62[i1][i2],
        _ => -4, // Default penalty for unknown amino acids
    }
}

/// Calculate tcrdist-style position score
/// Formula: max(0, 4 - BLOSUM62[a][b])
fn position_score(aa1: u8, aa2: u8) -> i32 {
    let blosum_score = blosum62_score(aa1, aa2) as i32;
    std::cmp::max(0, 4 - blosum_score)
}

/// Needleman-Wunsch alignment with tcrdist-style scoring
/// Returns alignment score (distance)
fn align_sequences(seq1: &str, seq2: &str, gap_penalty: i32) -> i32 {
    let seq1_bytes = seq1.as_bytes();
    let seq2_bytes = seq2.as_bytes();
    let len1 = seq1_bytes.len();
    let len2 = seq2_bytes.len();

    // Handle empty sequences
    if len1 == 0 && len2 == 0 {
        return 0;
    }
    if len1 == 0 {
        return (len2 as i32) * gap_penalty;
    }
    if len2 == 0 {
        return (len1 as i32) * gap_penalty;
    }

    // Initialize DP matrix
    let mut dp = vec![vec![0i32; len2 + 1]; len1 + 1];

    // Initialize first row and column
    for i in 0..=len1 {
        dp[i][0] = (i as i32) * gap_penalty;
    }
    for j in 0..=len2 {
        dp[0][j] = (j as i32) * gap_penalty;
    }

    // Fill DP matrix
    for i in 1..=len1 {
        for j in 1..=len2 {
            let match_score = dp[i - 1][j - 1] + position_score(seq1_bytes[i - 1], seq2_bytes[j - 1]);
            let delete_score = dp[i - 1][j] + gap_penalty;
            let insert_score = dp[i][j - 1] + gap_penalty;

            dp[i][j] = std::cmp::min(match_score, std::cmp::min(delete_score, insert_score));
        }
    }

    dp[len1][len2]
}

/// T-cell receptor with alpha and beta chain CDR sequences
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TCR {
    pub cdr1_a_aa: Option<String>,
    pub cdr2_a_aa: Option<String>,
    pub cdr3_a_aa: Option<String>,
    pub cdr1_b_aa: Option<String>,
    pub cdr2_b_aa: Option<String>,
    pub cdr3_b_aa: Option<String>,
}

impl TCR {
    pub fn new(
        cdr1_a_aa: Option<String>,
        cdr2_a_aa: Option<String>,
        cdr3_a_aa: Option<String>,
        cdr1_b_aa: Option<String>,
        cdr2_b_aa: Option<String>,
        cdr3_b_aa: Option<String>,
    ) -> Self {
        Self {
            cdr1_a_aa,
            cdr2_a_aa,
            cdr3_a_aa,
            cdr1_b_aa,
            cdr2_b_aa,
            cdr3_b_aa,
        }
    }
}

/// Calculate tcrdist distance between two TCRs
/// Combines alpha and beta chain distances
pub fn tcrdist(tcr1: &TCR, tcr2: &TCR) -> f64 {
    let alpha_dist = chain_distance(
        &tcr1.cdr1_a_aa,
        &tcr1.cdr2_a_aa,
        &tcr1.cdr3_a_aa,
        &tcr2.cdr1_a_aa,
        &tcr2.cdr2_a_aa,
        &tcr2.cdr3_a_aa,
    );

    let beta_dist = chain_distance(
        &tcr1.cdr1_b_aa,
        &tcr1.cdr2_b_aa,
        &tcr1.cdr3_b_aa,
        &tcr2.cdr1_b_aa,
        &tcr2.cdr2_b_aa,
        &tcr2.cdr3_b_aa,
    );

    (alpha_dist + beta_dist) as f64
}

/// Calculate distance for a single chain (alpha or beta)
fn chain_distance(
    cdr1_1: &Option<String>,
    cdr2_1: &Option<String>,
    cdr3_1: &Option<String>,
    cdr1_2: &Option<String>,
    cdr2_2: &Option<String>,
    cdr3_2: &Option<String>,
) -> i32 {
    let mut total_distance = 0;

    // CDR1 distance (weight = 1, gap penalty = 4)
    if let (Some(seq1), Some(seq2)) = (cdr1_1, cdr1_2) {
        total_distance += align_sequences(seq1, seq2, 4);
    }

    // CDR2 distance (weight = 1, gap penalty = 4)
    if let (Some(seq1), Some(seq2)) = (cdr2_1, cdr2_2) {
        total_distance += align_sequences(seq1, seq2, 4);
    }

    // CDR3 distance (weight = 3, gap penalty = 8)
    if let (Some(seq1), Some(seq2)) = (cdr3_1, cdr3_2) {
        total_distance += 3 * align_sequences(seq1, seq2, 8);
    }

    total_distance
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_blosum62_score() {
        // Test identical amino acids
        assert_eq!(blosum62_score(b'A', b'A'), 4);
        assert_eq!(blosum62_score(b'W', b'W'), 11);

        // Test different amino acids
        assert_eq!(blosum62_score(b'A', b'R'), -1);
        assert_eq!(blosum62_score(b'C', b'C'), 9);
    }

    #[test]
    fn test_position_score() {
        // Identical amino acids should have score 0
        assert_eq!(position_score(b'A', b'A'), 0);
        assert_eq!(position_score(b'W', b'W'), 0);

        // Different amino acids
        assert_eq!(position_score(b'A', b'R'), std::cmp::max(0, 4 - (-1)));
    }

    #[test]
    fn test_align_identical_sequences() {
        let score = align_sequences("CASSF", "CASSF", 4);
        assert_eq!(score, 0); // Identical sequences should have distance 0
    }

    #[test]
    fn test_align_different_sequences() {
        let score = align_sequences("CASS", "CASF", 4);
        assert!(score > 0); // Different sequences should have distance > 0
    }

    #[test]
    fn test_tcrdist_identical() {
        let tcr1 = TCR::new(
            Some("TGTGC".to_string()),
            Some("TGTGC".to_string()),
            Some("CASSF".to_string()),
            Some("TGTGC".to_string()),
            Some("TGTGC".to_string()),
            Some("CASSF".to_string()),
        );

        let dist = tcrdist(&tcr1, &tcr1);
        assert_eq!(dist, 0.0);
    }

    #[test]
    fn test_tcrdist_different() {
        let tcr1 = TCR::new(
            Some("TGTGC".to_string()),
            Some("TGTGC".to_string()),
            Some("CASSF".to_string()),
            Some("TGTGC".to_string()),
            Some("TGTGC".to_string()),
            Some("CASSF".to_string()),
        );

        let tcr2 = TCR::new(
            Some("TGTGA".to_string()),
            Some("TGTGA".to_string()),
            Some("CASSLF".to_string()),
            Some("TGTGA".to_string()),
            Some("TGTGA".to_string()),
            Some("CASSLF".to_string()),
        );

        let dist = tcrdist(&tcr1, &tcr2);
        assert!(dist > 0.0);
    }
}
